e2e/playwright/conftest.py

```diff
@@ -0,0 +1,419 @@
+
+        return stdout
+
+    def __enter__(self):
+        self.start()
+        return self
+
+    def start(self):
+        # Start the process and capture its stdout/stderr output to a temp
+        # file. We do this instead of using subprocess.PIPE (which causes the
+        # Popen object to capture the output to its own internal buffer),
+        # because large amounts of output can cause it to deadlock.
+        self._stdout_file = TemporaryFile("w+")
+        print(f"Running: {shlex.join(self.args)}")
+        self._proc = subprocess.Popen(
+            self.args,
+            cwd=self.cwd,
+            stdout=self._stdout_file,
+            stderr=subprocess.STDOUT,
+            text=True,
+            env={**os.environ.copy(), **self.env} if self.env else None,
+        )
+
+    def __exit__(self, exc_type, exc_val, exc_tb):
+        if self._proc is not None:
+            self._proc.terminate()
+            self._proc = None
+        if self._stdout_file is not None:
+            self._stdout_file.close()
+            self._stdout_file = None
+
+
+def resolve_test_to_script(test_module: ModuleType) -> str:
+    """Resolve the test module to the corresponding test script filename."""
+    module_name = test_module.__name__
+    return module_name.replace("_test", ".py")
+
+
+def find_available_port(host: str = "localhost") -> int:
+    """Find an available port on the given host."""
+    with contextlib.closing(socket.socket(socket.AF_INET, socket.SOCK_STREAM)) as s:
+        s.bind((host, 0))  # 0 means that the OS chooses a random port
+        s.setsockopt(socket.SOL_SOCKET, socket.SO_REUSEADDR, 1)
+        return int(s.getsockname()[1])
+
+
+def is_app_server_running(port: int, host: str = "localhost") -> bool:
+    """Check if the app server is running."""
+    try:
+        return (
+            requests.get(f"http://{host}:{port}/_stcore/health", timeout=1).text == "ok"
+        )
+    except Exception:
+        return False
+
+
+def wait_for_app_server_to_start(port: int, timeout: int = 5) -> bool:
+    """Wait for the app server to start.
+
+    Parameters
+    ----------
+    port : int
+        The port on which the app server is running.
+
+    timeout : int
+        The number of minutes to wait for the app server to start.
+
+    Returns
+    -------
+    bool
+        True if the app server is started, False otherwise.
+    """
+
+    print(f"Waiting for app to start... {port}")
+    start_time = time.time()
+    while not is_app_server_running(port):
+        time.sleep(3)
+        if time.time() - start_time > 60 * timeout:
+            return False
+    return True
+
+
+def wait_for_app_run(page: Page):
+    """Wait for the app to finish running."""
+    # Wait until the script is no longer running.
+    page.wait_for_selector(
+        "[data-testid='stStatusWidget']", timeout=20000, state="detached"
+    )
+    # Give the app a little more time to render everything
+    page.wait_for_timeout(500)
```

Comment:
this feels flakey / not best practice. Is there something else to wait for?

Reply:
Yep, indeed. It's not the best practice. I removed it for now, but removing it might actually add some more risk of flakiness. But lets see have stable the tests will be without any waiting.